File: SearchingAndSorting/FourNumberSum/FourNumberSum.js
function method1(arr, n, target) {
  let i,
    j,
    k,
    l,
    ind = 0;
  let sum = 0;
  let res = [];
  for (i = 0; i < n - 3; i++) {
    for (j = i + 1; j < n - 2; j++) {
      for (k = j + 1; k < n - 1; k++) {
        for (l = k + 1; l < n; l++) {
          sum = arr[i] + arr[j] + arr[k] + arr[l];
          if (sum == target) {
            res[ind++] = [arr[i], arr[j], arr[k], arr[l]];
          }
        }
      }
    }
  }
  res.sort();
  return res;
}



function method2(arr, n, target) {
  let i,
    j,
    k,
    l,
    temp = 0,
    ind = 0;
  let res = {};
  arr.sort((a,b)=>parseInt(a)-parseInt(b));
  console.log('arr is ', arr);
  for (i = 0; i < n-3; i++) {
    for (j = i + 1; j < n-2; j++) {
      k = j + 1;
      l = n - 1;
      while (k < l) {
          console.log("i : ",i,", j : ",j,", k : ",k,", l : ",l);
        temp = arr[i] + arr[j] + arr[k] + arr[l];
        if (temp == target) {
          let str = ''+arr[i]+arr[j]+arr[k]+arr[l];
          console.log('str is ', str);
          res[str] = [arr[i], arr[j], arr[k], arr[l]];
          k++;
          l--;
        }
        else if(temp>target){
            l--;
        }else{
            k++;
        }
      }
    }
  }
  return Object.values(res);
}

let arr = [-1, -2, -2, -2, -4, -5,  0,  0,  1]
let target = -9;
//console.log(method1(arr, arr.length, target));
console.log(method2(arr, arr.length, target));

File: SearchingAndSorting/FourNumberSum/FourNumberSum.go
package main

import (
	"fmt"
	"sort"
)

func method1(arr []int, sum int) [4]int {
	res := [4]int{}
	n := len(arr)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			for k := j + 1; k < n; k++ {
				for l := k + 1; l < n; l++ {
					if arr[i]+arr[j]+arr[k]+arr[l] == sum {
						res[0] = arr[i]
						res[1] = arr[j]
						res[2] = arr[k]
						res[3] = arr[l]
						return res
					}
				}
			}
		}
	}
	return res
}

func method2(arr []int, sum int) [4]int {
	res := [4]int{}
	sort.Ints(arr)
	n := len(arr)

	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			l := j + 1
			r := n - 1
			temp := arr[i] + arr[j]
			for l < r {
				cur := arr[l] + arr[r]
				if temp+cur > sum {
					r--
				} else if temp+cur < sum {
					l++
				} else {
					res[0] = arr[i]
					res[1] = arr[j]
					res[2] = arr[l]
					res[3] = arr[r]
					return res
				}
			}
		}
	}
	return res
}

func main() {
	arr := []int{-1, -2, -2, -2, -4, -5, 0, 0, 1}
	sum := -9
	fmt.Println(method1(arr, sum))
	fmt.Println(method2(arr, sum))
}
